waspc/src/Wasp/SemanticVersion.hs

```diff
@@ -18,24 +27,51 @@ data Version = Version
 instance Show Version where
   show (Version mjr mnr ptc) = printf "%d.%d.%d" mjr mnr ptc
 
-data VersionBounds
-  = -- | Allows only the version exactly equal to the one specified in the bounds
-    Exact Version
-  | -- | Allows changes that do not modify the leftmost non-zero digit in major.minor.patch, as described
-    -- in node semver docs: https://github.com/npm/node-semver#caret-ranges-123-025-004
-    BackwardsCompatibleWith Version
+data Operator = Equal | LessThanOrEqual | BackwardsCompatibleWith
+
+instance Show Operator where
+  show Equal = "="
+  show LessThanOrEqual = "<="
+  show BackwardsCompatibleWith = "^"
+
+data Comparator = Comparator Operator Version
+
+instance Show Comparator where
+  show (Comparator op v) = show op ++ show v
```

Comment:
Awesome analysis @sodic, thanks for this!

From their spec, I got the idea that they use Version - Operator - Comparator - ComparatorSet - Range as a basic structure, and then the rest is sugar on top of it (they do call it "sugar" actually).
So I followed the same approach, and decided this will be the basis, and we will add other special stuff, like special ranges, comparators and operators on top of that, somehow.
I did the exception of adding `^` directly as the primitive operator though, because I saw no reason why that wouldn't work, and I didn't want to further complicate things for now. But I agree, we should take that one out, since it is a special comparator set that desugarizes into the primitive comparators.

I think you are voting for (1) :D, but I would suggest we go with (3) + a TODO for `^`, since that will allow us to merge it right now + I think it is a good foundation for further expanding this code as needed. While I agree it might be a bit overly complex, a big benefit is that we can just copy what they did without making up our stuff. We could reduce the complexity, but then when we decide to add something we might uncover that we need more structure and will likely again end up with what they did, or with something else that has different names and maybe structure but is similarly complex, assuming their solution is relatively good / not far away from minimal required complexity.

Reply:
Makes sense, we merged it and agreed on future steps (written in todos).